쥬스메이커 [Step02] 송준, Rowan

kimseongj opened the pull request:
@wongbingg
안녕하세요 웡빙~ step2 PR 보내드립니다.
이번 코드는 다소 난잡해 보이지만 리뷰 잘부탁드립니다😂

---

## 고민한 점 
### ViewController의 ```fruitsStock``` 프로퍼티
- 값 타입? 참조 타입?
요구조건 "과일쥬스를 제조하여 과일의 재고가 변경되면 화면의 적절한 요소에 변경사항을 반영합니다." 를 구현할 때 문제점이 있었습니다.

```swift
// 수정 전
final class ViewController: UIViewController {
    private let juiceMaker = JuiceMaker()
    private var fruitsStock = FruitStore.shared.fruitsStock
    
    override func viewDidLoad() {
        super.viewDidLoad()
        displayStock()
    }
    
    func displayStock() {
        if let strawberryStock = fruitsStock[.strawberry],
           ... {
            stockOfStrawberry.text = String(strawberryStock)
            ...
        }
    }
    
    @IBAction func orderStrawberryBananaJuice(_ sender: UIButton) {
        do {
            try juiceMaker.make(juice: .strawberryBanana)
            displayStock()
        } catch ... 
    }
}
```
>```fruitsStock``` 변수에 ```FruitStore.shared.fruitsStock```을 할당했을 경우 화면에 변경사항이 반영되지 않았습니다.
값 타입의 특성을 잘 몰랐기 때문에 발생한 문제였다고 생각합니다.
>
>재고 초기값이 10인 상태로 ```fruitsStock```변수에 초기값이 복사되어있어 ```make```메서드를 통해 재고가 변경되어도 화면에는 그대로 10이 출력되었습니다.
>
>Swift의 Copy on Write에 대한 학습에 더불어 월요일 학습활동을 통해 값 타입과 참조 타입의 차이점을 알게 되었습니다.

```swift
// 수정 후
final class ViewController: UIViewController {
    private let juiceMaker = JuiceMaker()
    private var fruitsStock: [Fruits: Int] {
        return FruitStore.shared.fruitsStock
    } 
    
    override func viewDidLoad() {
        super.viewDidLoad()
        displayStock()
    }
    
    func displayStock() {
        if let strawberryStock = fruitsStock[.strawberry],
           ... {
            stockOfStrawberry.text = String(strawberryStock)
            ...
        }
    }
    
    @IBAction func orderStrawberryBananaJuice(_ sender: UIButton) {
        do {
            try juiceMaker.make(juice: .strawberryBanana)
            displayStock()
        } catch ... 
    }
}
```
>```fruitsStock```을 연산 프로퍼티로 정의하여 해당 변수에 접근하는 시점에 재고 딕셔너리 값을 가져올 수 있도록 만들어 해결했습니다.

<br/>

## 해결이 되지 않은 점 
재고수정과, Alert에서 동일한 화면으로 전환시키기 위해 똑같은 방법을 사용하였습니다. NavigationController에서 Push해주는 방법을 사용했을 때, 재고수정버튼을 누를 경우 오른쪽에서 슬라이딩되며 화면전환이 되는데, Alert을 통해 화면전환을 할 때는 부자연스럽게 화면전환이 됩니다.
어떤 문제점이 있는걸까요?

- 재고수정버튼을 통한 화면 이동
```swift
@IBAction func ClickChangeStock(_ sender: Any) {
        moveToChangeStockViewController()
    }
```
![재고수정버튼-화면이동](https://user-images.githubusercontent.com/88870642/211257941-5e928bef-a47d-4283-8ac2-e3fd9a87e1f9.gif)

<br/>

- Alert을 통한 화면 이동
```swift
func setFailAlert() {
        let failAlert = UIAlertController(title: "재료가 모자라요. 재고를 수정할까요?", message: nil, preferredStyle: UIAlertController.Style.alert)
        let yesAction = UIAlertAction(title: "예",
                                      style: UIAlertAction.Style.default,
                                      handler: { Action in self.moveToChangeStockViewController() })
        
        let noAction = UIAlertAction(title: "아니오", style: UIAlertAction.Style.default, handler: nil)
        failAlert.addAction(yesAction)
        failAlert.addAction(noAction)
        present(failAlert, animated: true, completion: nil)
    }
```
![alert-화면이동](https://user-images.githubusercontent.com/88870642/211257949-104079b5-b576-48f2-a33d-1d371a8b87c4.gif)

<br/>

## 조언을 얻고 싶은 점
### 쥬스 주문 버튼의 IBAction에서 반복되는 코드를 줄이는 방법
@IBAction 함수에서 매개변수를 받는 방법은 없는건지 궁금합니다! 🧐
밑에 코드에서 ```.strawberryBanana```를 받아올 수 있다면 ```do-catch```구문을 함수로 뺄 수 있을거 같습니다. 
```swift=
@IBAction func orderStrawberryBananaJuice(_ sender: UIButton) {
        do {
            try juiceMaker.make(juice: .strawberryBanana)
            setSuccessAlert(juiceName: .strawberryBanana)
            displayStock()
        } catch StockError.outOfStock {
            setFailAlert()
        } catch {
            print(error)
        }
    }
```
### 연산 프로퍼티 작동원리가 궁금합니다!
getter가 작동을 할 때, return값을 복사해서 ```fruitsStock```에 할당하는건지, 아니면 ```fruitsStock```에 할당되지 않고 return값 자체만 반환하는건지 궁금합니다!
 ```swift 
private var fruitsStock: [Fruits: Int] {
        return FruitStore.shared.fruitsStock 
    }
 ```

Tediousday93 committed:
feat: 재고변경 화면 전환

kimseongj committed:
feat: ViewController, ChangeStockViewController 클래스 생성

Tediousday93 committed:
refactor: FruitStore - subtractStock, checkStock 메서드 이름 변경 및 코드 수정

Tediousday93 committed:
refactor: make, isEnoughStock 메서드 코드 수정 (throw 타이밍 조절)

kimseongj committed:
feat: 주문완료 Alert 생성

kimseongj committed:
docs: README파일 Step1 작성

Tediousday93 committed:
docs: Update README

오타 수정

Tediousday93 committed:
feat: failAlert 구현, JuiceMaker Juice열거형 name 연산 프로퍼티 추가

kimseongj committed:
feat: FruitStore 싱글톤 객체 생성,  displayStock 함수 생성

Tediousday93 committed:
feat: displayStock 메서드 딸기 외 나머지 과일 텍스트 추가

kimseongj committed:
chore: 주석정리

wongbingg commented:
안녕하세요 송준, Rowan 쥬스메이커 프로젝트 step2 고생하셨습니다 ☺️

>저의 리뷰는 여러분과 함께 고민하는 과정입니다. 제가 드리는 피드백이 정답은 아니며, 여러분의 선택이 정답이 됩니다. 더 넓게 생각하는 자세를 가지고 제 피드백을 봐주셨으면 좋겠습니다 ☺️

### ✅ 고민한 점

먼저 사용한 방법은 `FruitStore.shared.fruitsStock` 으로 접근하는 것을 보니 싱글톤 패턴을 사용하셨네요~! 싱글톤 패턴으로 사용하는 객체는 어떤 장점과 단점을 가지고 있을까요? 여기서 FruitStore를 싱글톤 패턴으로 사용했을 때와 사용하지 않았을 때의 차이점은 무엇일까요? 

### ✅ 해결이 되지 않은 점

moveToChangeStockViewController() 메서드에서 화면전환 메서드인 pushViewController() 가 사용되고 있는데요, 이 메서드에서 설정해줄 수 있을 것 같아요! 파라미터를 잘 살펴볼까요 ?

### ✅ 조언을 얻고 싶은 점 

#### 쥬스 주문 버튼의 IBAction에서 반복되는 코드를 줄이는 방법
- @IBAction 하나에 여러개의 버튼을 연결해줄 수 있다는 점을 알고 계셨나요? 또한 현재 파라미터로 받고있는 sender를 활용해볼 수 있을 것 같아요! sender 는 UIButton 타입이니 어떤 버튼을 눌렀는지 구분해줄 수 있는 UIButton의 프로퍼티 값이 있다면 각 버튼으로 분기처리를 해줄 수 있을 것 같아요.

#### 연산 프로퍼티의 작동 원리가 궁금합니다!
- [Swift Programming Language - Properties](https://docs.swift.org/swift-book/LanguageGuide/Properties.html)

- 위 공식문서에서 **Computed Propertie** 부분을 보시면 첫 번째 문단에서 해당 내용에 대한 답이 나옵니다 ☺️

wongbingg commented on JuiceMaker/JuiceMaker/Controller/ChangeStockViewController.swift:
```diff
@@ -0,0 +1,46 @@
+//  JuiceMaker - ChangeStockViewController.swift
+//  Created by 송준, Rowan on 2023/01/05.
```
주석도 꼼꼼하게 변경하는 습관 좋습니다 👍

wongbingg commented on JuiceMaker/JuiceMaker/Controller/ChangeStockViewController.swift:
```diff
@@ -0,0 +1,46 @@
+import UIKit
+
+class ChangeStockViewController: UIViewController {
+    
```
이 클래스가 상속될 가능성이 있을까요 ? 상속될 가능성이 없다면 붙여줄 수 있는 키워드가 있을 것 같아요 🧐

Tediousday93 replied:
```final``` 키워드를 붙여서 상속을 방지했습니다!

kimseongj replied:
final 클래스로 변경 완료했습니다!
d68ec8f21d1893a65e7db465828ed78473a9c0b1

wongbingg commented on JuiceMaker/JuiceMaker/Controller/ChangeStockViewController.swift:
```diff
@@ -0,0 +1,46 @@
+    @IBOutlet weak var numberOfKiwi: UILabel!
+    @IBOutlet weak var numberOfMango: UILabel!
+    
+    @IBAction func pushPlusStrawberry(_ sender: Any) {
+        //
+    }
+    @IBAction func pushMinusStrawberry(_ sender: Any) {
+        //
+    }
+    @IBAction func pushPlusBanana(_ sender: Any) {
+        //
+    }
+    @IBAction func pushMinusBanana(_ sender: Any) {
+        //
+    }
+    @IBAction func pushPlusPineApple(_ sender: Any) {
+        //
+    }
+    @IBAction func pushMinusPineApple(_ sender: Any) {
+        //
+    }
+    @IBAction func pushPlusKiwi(_ sender: Any) {
+        //
+    }
+    @IBAction func pushMinusKiwi(_ sender: Any) {
+        //
+    }
+    @IBAction func pushPlusMango(_ sender: Any) {
+        //
+    }
+    @IBAction func pushMinusMango(_ sender: Any) {
+        //
+    }
```
sender 파라미터의 타입이 Any로 되어있네요. 구체적인 타입으로 받아주면 처리하기 더 수월해질 것 같아요.
이 부분은 다음 스텝에서 더 코멘트를 남기도록 할게요 🧐

wongbingg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,17 +1,149 @@
 //  JuiceMaker - ViewController.swift
-//  Created by yagom. 
+//  Created by yagom.
```
여기 주석은 Rowan, 송준으로 되어있지 않네요 !? 🧐
주석도 컨벤션을 유지해주면 좋을 것 같아요

kimseongj replied:
주석 컨벤션 수정했습니다!

wongbingg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,17 +1,149 @@
+        self.navigationController?.pushViewController(nextVC, animated: false)
+    }
+    
+    func setSuccessAlert(juiceName: JuiceMaker.Juice) {
+        let successAlert = UIAlertController(title: "\(juiceName.name)가 나왔습니다 맛있게 드세요", message: nil, preferredStyle: UIAlertController.Style.alert)
+        let offAction = UIAlertAction(title: "확인", style: UIAlertAction.Style.default, handler: nil)
+        successAlert.addAction(offAction)
+        present(successAlert, animated: true, completion: nil)
+    }
```
```suggestion
    func setSuccessAlert(juiceName: JuiceMaker.Juice) {
        let successAlert = UIAlertController(
            title: "\(juiceName.name)가 나왔습니다 맛있게 드세요",
            message: nil, 
            preferredStyle: UIAlertController.Style.alert
        )
        let offAction = UIAlertAction(
            title: "확인", 
            style: UIAlertAction.Style.default,
            handler: nil
        )
        successAlert.addAction(offAction)
        present(successAlert, animated: true, completion: nil)
    }
```

보통 파라미터가 3개 이상일 때는 이런 식으로 개행을 해주기도 합니다 ☺️ 스타일이니 참고만 해주세요

wongbingg replied:
파라미터명이 juiceName이라 사용할 때 juiceName.name 으로 나오는 것이 어색한 것 같아요 파라미터 명을 수정해볼까요?

kimseongj replied:
```juiceName.name``` 대신 ```juice.name```으로 변경했습니다!

Tediousday93 replied:
name의 중복 사용을 피하기 위해 파라미터를 ```juice```로 변경했습니다!
title, message, preferredStyle도 개행을 적용해봤어요🧐
개행하는게 가독성이 더 좋아보이네요!!

wongbingg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,17 +1,149 @@
+    }
+    
+    @IBAction func orderMangoKiwiJuice(_ sender: UIButton) {
+        do {
+            try juiceMaker.make(juice: .mangoKiwi)
+            setSuccessAlert(juiceName: .mangoKiwi)
+            displayStock()
+        } catch StockError.outOfStock {
+            setFailAlert()
+        } catch {
+            print(error)
+        }
```
위에 코멘트에서 말씀 드렸던것 처럼 sender 를 활용해보면 반복되는 7개의 메서드를 줄여줄 수 있을 것 같아요 🧐

kimseongj replied:
각각의 버튼들을 IBOutlet을 통해 선언하고, button에 따라 Switch문을 통해 Juice타입으로 반환시켜줬습니다!
786d662abde808b50ab685c7ffccf51a304729f2

wongbingg replied:
코드가 간단해졌군요 ! 👍

wongbingg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -12,29 +13,27 @@ class FruitStore {
-            throw StockError.outOfStock
+    func subtractStock(of fruit: Fruits, count number: Int) {
+        if let selectedStock = FruitStore.shared.fruitsStock[fruit] {
+            fruitsStock[fruit] = selectedStock - number
         }
-        
-        fruitsStock[fruit] = selectedStock - number
     }
```
이 함수가 사용되고 있는 타입 내에 바로 접근이 가능한 `fruitsStock` 프로퍼티가 있음에도 불구하고 FruitStore.shared.fruitsStock으로 접근하는 것이 뭔가 어색한 것 같아요 !  `FruitStore.shared.fruitsStock` -> `fuitStock` 이렇게 변경하면 문제가 생길까요 ?

wongbingg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -12,29 +13,27 @@ class FruitStore {
+        if let selectedStock = FruitStore.shared.fruitsStock[fruit] {
             fruitsStock[fruit] = selectedStock + number
         }
     }
     
-    func checkStock(of fruit: Fruits) {
-        if let stock = fruitsStock[fruit] {
-            let resultMessage = "\(fruit)의 재고는 \(stock)개입니다."
-            print(resultMessage)
+    func isEnoughStock(of fruit: Fruits, count number: Int) -> Bool {
+        guard let selectedStock = FruitStore.shared.fruitsStock[fruit],
+              selectedStock - number >= 0 else {
+            return false
         }
+        return true
     }
```
나머지 부분도 마찬가지로 `fruitsStock` 으로 접근 가능한 부분은 shared 를 통해 접근하는 것으로 변경 해주셨네요. 어떤 이유로 이 부분을 변경 해주셨을까요?

Tediousday93 replied:
싱글톤 인스턴스를 만들면서 혼동한 것 같네요 😅
```FruitStore.shared.fruitsStock```을 ```fruitsStock```으로 수정했습니다!

wongbingg commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -20,21 +20,33 @@ struct JuiceMaker {
             }
         }
+        
+        var name: String {
+            switch self {
+            case .mango: return "망고쥬스"
+            case .pineapple: return "파인애플쥬스"
+            case .banana: return "바나나쥬스"
+            case .kiwi: return "키위쥬스"
+            case .strawberry: return "딸기쥬스"
+            case .strawberryBanana: return "딸바쥬스"
+            case .mangoKiwi: return "망키쥬스"
+            }
+        }
```
계산프로퍼티로 깔끔하게 잘 분리 해주셨네요 👍

wongbingg commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -20,21 +20,33 @@ struct JuiceMaker {
-                fruitStorage.checkStock(of: key)
+        for (key, value) in ingredients {
+            guard fruitStorage.isEnoughStock(of: key, count: value) else {
+                throw StockError.outOfStock
             }
-            print("주문하신 \(juice.rawValue)가 나왔습니다!")
-        } catch StockError.outOfStock {
-            print("재고가 부족합니다.")
-        } catch {
-            print(error)
+        }
+
+        for (key, value) in ingredients {
+            fruitStorage.subtractStock(of: key, count: value)
+        
+        print("주문하신 \(juice.name)가 나왔습니다!")
         }
     }
 }
```
make() 함수에 대한 오류처리를 잘 해주셨네요 👍

wongbingg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,17 +1,149 @@
+        displayStock()
+    }
+    
+    @IBOutlet weak var stockOfStrawberry: UILabel!
+    @IBOutlet weak var stockOfBanana: UILabel!
+    @IBOutlet weak var stockOfPineApple: UILabel!
+    @IBOutlet weak var stockOfKiwi: UILabel!
+    @IBOutlet weak var stockOfMango: UILabel!
```
@IBOutlet 프로퍼티는 `weak` 키워드와 `Implicitly Unwrapped Optional`을 사용하고 있는데요, **시간이 남는다면 !** 왜 이렇게 사용되는지에 대해 알아보셔도 좋을 것 같아요

kimseongj replied:
#### ```Implicitly Unwrapped Optional```를 사용하는 이유
1. 애플이 권장하는 방식입니다. (자동완성)
2. Swift 언어 특성상, ```struct```나 ```class```가 인스턴스로 생성될 때 저장 프로퍼티에 초기값을 저장해야 합니다. 그렇기 때문에 Viewcontroller가 인스턴스화될 때 IBOutlet의 값은 항상 존재해야 합니다. 고로 강제 언래핑을 사용을 추천하는거 같습니다.

#### 참고 
- [Optional IBOutlet Blog1](https://blog.curtisherbert.com/to-optional-or-not-to-optional-iboutlet/)
- [Optional IBOutlet Blog2](https://cocoacasts.com/should-outlets-be-optionals-or-implicitly-unwrapped-optionals)

Tediousday93 replied:
```weak```키워드 사용 이유에 대해 조사해봤습니다...💭

### Strong / Weak Reference
프로젝트를 진행하면서 ```@IBOutlet```을 정의할 때마다 ```weak var```를 사용했다. Outlet에 weak 참조를 사용하는 이유는 뭘까?
먼저 Swift의 ARC를 통한 Memory Deallocation 및 Strong Reference와 Weak Reference의 차이를 알아야 한다.

- **Memory Deallocation**
    - swift에서 메모리에 올라간 데이터는 ARC(Auto Reference Count)에 의해 자동으로 참조 횟수가 조절되며 Reference Count가 0이 되면 메모리에서 할당 해제된다.
    - property의 Reference type default는 ```strong```이다.
    - 임의로 인스턴스를 deallocation하려면 ```nil```을 할당한다.

- **Strong Reference**
    - 강한 참조: Reference Count가 참조가 발생할 때마다 증가함 -> 복수 참조가 일어났을 경우, Reference Count가 1보다 커져 nil을 할당하여 Reference Count를 1 감소시키더라도 메모리에서 할당 해제되지 않음(인스턴스가 없어져도 메모리에서 Deallocation되지 않음)

- **Weak Reference**
    - 약한 참조: Reference Count가 1을 초과하지 않음 -> Deallocation 방지 X (nil을 할당하면 메모리에서 Deallocation됨)
    - 객체가 nil일 경우가 존재하므로 항상 옵셔널 타입이어야 한다. (```UIButton!```)
    - ARC에 의해 자동으로 Deallocation됨!

따라서, IBOutlet에 weak참조를 사용하는 것은 viewDidLoad 이전에 outlet에 연결된 view들이 nil일 수 있고 아직까지는 view hierarchy가 복잡하지 않아 view의 deallocation을 방지할 이유가 없기 때문이라고 생각한다..

[참조]
1. [프로그래미즈 - Swift Strong and Weak References](https://www.programiz.com/swift-programming/strong-weak-reference)
2. [srkim blog - 메모리를 참조하는 방법](https://devsrkim.tistory.com/entry/Swift-%EB%A9%94%EB%AA%A8%EB%A6%AC%EB%A5%BC-%EC%B0%B8%EC%A1%B0%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95-Strong-Weak-Unowned)

wongbingg replied:
정말 자세하게 알아보셨군요 대단합니다 👍 지금 이해하지 못해도 괜찮은 내용들입니다 ! 이 부분도 캠프 중반부 쯤 배우게 될 내용이에요~! 미리 한번 훑어보면 좋을 것 같아서 말씀 드렸었습니다 ㅎㅎ

kimseongj committed:
chore: 주석정리

Tediousday93 committed:
feat: NameSpace AlertMessege 생성, ChangeStockVC 상속 가능성을 없애기 위한 final keyword 추가

kimseongj committed:
chore: 부가적인 변수명 삭제, 필요 없는 코드 삭제

FruitStore클래스에서 subtractStock, addStock, isEnoughStock 메서드에서 "FruitStore.shared.fruitsStock[fruit]" -> "fruitsStock[fruit]" 변경
ViewController클래스에서 필요 없는 코드 삭제

Tediousday93 committed:
refactor: order***Juice 메서드를 pushOrderButton으로 통합

통합에 따른 쥬스 주문 버튼 IBOutlet 생성

Tediousday93 commented:
@wongbingg 
항상 추가로 학습해볼 내용을 던져주셔서 보람찬 하루를 보내고 있습니다 🔥🔥🔥
피드백 감사합니다!!

## 피드백 이후 refactoring 사항
### NameSpace
```swift
enum AlertMessege {
        static let confirm = "확인"
        static let yes = "예"
        static let no = "아니오"
        static let madeJuice = "가 나왔습니다. 맛있게 드세요!"
        static let lackOfStock = "재료가 모자라요. 재고를 수정할까요?"
        static let success = "성공"
        static let failure = "실패"
    }
```
Alert Messege 네임스페이스를 정의하여 매직 리터럴 사용을 줄여보았습니다!

---

### orderJuice 메서드 수정
```swift
@IBAction func pushOrderButton(_ sender: UIButton) {
        guard let selectedJuice = identifyJuice(of: sender) else { return }
        
        order(selectedJuice)
    }
    
    func identifyJuice(of button: UIButton) -> JuiceMaker.Juice? {
        switch button {
        case orderStrawberry: return .strawberry
        case orderBanana: return .banana
        case orderPineapple: return .pineapple
        case orderKiwi: return .kiwi
        case orderMango: return .mango
        case orderStrawberryBanana: return .strawberryBanana
        case orderMangoKiwi: return .mangoKiwi
        default: return nil
        }
    }
    
    func order(_ juice: JuiceMaker.Juice) {
        do {
            try juiceMaker.make(juice: juice)
            setSuccessAlert(juice: juice)
            displayStock()
        } catch StockError.outOfStock {
            setFailAlert()
        } catch {
            print(error)
        }
    }
```

 모든 버튼을 하나의 ```@IBAction orderJuice```에 연결하면서 메서드 이름도 적절하게 바꿔주고 함수 분리도 해봤습니다. 훨씬 보기 좋아진 것 같아요 👍
 
```identifyJuice```메서드의 경우, ```UIButton.tag```를 사용해볼까 고민했는데 결국 ```Int``` 값을 가지는 tag를 사용하면 매직 넘버의 사용이 불가피할 것 같아 버튼의 ```IBOutlet```을 만들어 ```switch-case```문으로 버튼 식별 기능을 구현해봤습니다.

Brody, Goat 조 처럼 ```sender.titleLabel?.text```를 활용하는 방법은 미처 생각하지 못했네요. 이것도 좋은 방법이라고 생각합니다.

wongbingg commented:
코멘트에 대한 수정에 추가 리팩토링 까지 👍 대단합니다 로웬과 송준 !!
저도 여러분의 열정을 본받아야 겠군요!  이번 스텝에서 피드백에 대한 수정이 잘 이루어져
머지하도록 하겠습니다 ☺️ 다음 스텝도 화이팅입니다 💪💪
